ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+
+static char *llite_stats_uint64_t_entries[] = {
+        "dirty_pages_hits",
+        "dirty_pages_misses",
+        "read_bytes.sum",
+        "write_bytes.sum",
+        "brw_read.sum",
+        "brw_write.sum",
+        "ioctl",
+        "open",
+        "close",
+        "mmap",
+        "page_fault",
+        "page_mkwrite",
+        "seek",
+        "fsync",
+        "readdir",
+        "setattr",
+        "truncate",
+        "flock",
+        "getattr",
+        "create",
+        "link",
+        "unlink",
+        "symlink",
+        "mkdir",
+        "rmdir",
+        "mknod",
+        "rename",
+        "statfs",
+        "alloc_inode",
+        "setxattr",
+        "getxattr",
+        "getxattr_hits",
+        "listxattr",
+        "removexattr",
+        "inode_permission",
+        NULL
+};
+
+/*******************************************************************/
+/* ldms_metrict_by_name() employs a linear lookup and is verboten for
+ * use by plugins in the main sample() loop.  This section section
+ * implemants a faster lookup using the rbt structure, and offers
+ * alt_ldms_metric_by_name() as a limited alternative to the original.
+ * Someday, perhaps ldms_metric_by_name() will be improved and this
+ * section can be removed.
+ */
+/* red-black tree root for metric index lookup table */
+static struct rbt metric_lookup_tree;
+struct metric_lookup_entry {
+        struct rbn node;
+        int index;
+};
+
+static int metric_lookup_tree_insert(const char *name, int index)
+{
+        struct metric_lookup_entry *entry;
+
+        entry = malloc(sizeof(struct metric_lookup_entry));
+        if (entry == NULL) {
+                return ENOMEM;
+        }
+        entry->index = index;
+        rbn_init(&entry->node, (char *)name);
+        rbt_ins(&metric_lookup_tree, &entry->node);
+
+        return 0;
+}
+
+static void metric_lookup_tree_destroy()
+{
+        struct rbn *rbn;
+        struct metric_lookup_entry *entry;
+
+        while (!rbt_empty(&metric_lookup_tree)) {
+                rbn = rbt_min(&metric_lookup_tree);
+                entry = container_of(rbn, struct metric_lookup_entry, node);
+                rbt_del(&metric_lookup_tree, rbn);
+                free(entry);
+        }
+}
+
+/* "set" is unused in this function.  It is just there to allow this
+   call to be a direct replacement for ldms_metric_by_name().  This plugin
+   only uses one schema, so there is only one global lookup table. */ 
+static int alt_ldms_metric_by_name(ldms_set_t set, const char *name)
+{
+        struct metric_lookup_entry *entry;
+        struct rbn *rbn;
+
+        rbn = rbt_find(&metric_lookup_tree, name);
+        if (rbn == NULL) {
+                return -1;
+        }
+        entry = container_of(rbn, struct metric_lookup_entry, node);
+
+        return entry->index;
+}
+/*******************************************************************/
+
+static int llite_general_schema_is_initialized()
```

Comment:
This function is unneeded. if we fix config() to create the schema. If config() cannot create the schema, then it returns an error and the plugin will never be sampled, so sample never needs to check for schema existence.

Reply:
config() is not broken.  This plugin is explicitly designed to require no configuration.  ldmsd does not require that config() be called before the plugin is started, and this plugin does not require that config() be called to work. It will operate just fine for many use cases by running on defaults.

And because ldmsd does not enforce that "config" is executed before "start" for a plugin, if I were to change the implementation as you suggest, I would just have to create a new "config_was_called()" check anyway to enforce that config() is called before sample().

I'm not suggesting that ldmsd should change.  I think it is entirely reasonable, and in fact desirable, to be able to start a plugin without needing to config() it if the plugin can reasonably operate in that mode.  Anything that reduces the configuration burden on sysadmins is desireable.

But the current implementation is intentional.

Reply:
Always calling config is 'standard practice' in v3 and is enforced in v4 via sampler_base::base_config
config values for producer and instance are always required, So config is effectively always required. conveniently, producer is in most cases (including lustre_client)  going to be given the same value as hostname {this differs for things like the router identifiers, though}.
There is no configuration burden on sysadmins at this level of detail/defaults if they use the systemd scripts shipped with ldms and based on libgenders.

Reply:
Use of "sampler_base::base_config" seems to be a voluntary policy enforcement rather than a code enforcement.  I think that the ldmsd plugin code really needs to add a flag to track whether or not a plugin has config() called (and also returned a success return code) before it allows "start" to be issued for that plugin.  If that is really the semantics, then check and enforce them so that we don't need to write a bunch of the same checks over and over again in every plugin.

Offering a library that does those checks isn't the same.  In fact, I already have issues with base_config not being designed in a way that works well with my plugins, so I didn't use them for v4.  But the plugin config/start semantics should still be enforced correctly by ldmsd.